src/DotNetty.Transport/Channels/Groups/DefaultChannelGroup.cs

```diff
@@ -0,0 +1,387 @@
+﻿using DotNetty.Buffers;
+using DotNetty.Common.Concurrency;
+using DotNetty.Common.Utilities;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace DotNetty.Transport.Channels.Groups
+{
+    public class DefaultChannelGroup : IChannelGroup
+    {
+        static int nextId = 0;
+        readonly string name;
+        readonly IEventExecutor executor;
+        readonly ConcurrentDictionary<IChannelId, IChannel> serverChannels = new ConcurrentDictionary<IChannelId, IChannel>();
+        readonly ConcurrentDictionary<IChannelId, IChannel> nonServerChannels = new ConcurrentDictionary<IChannelId, IChannel>();
+
+        public DefaultChannelGroup(IEventExecutor executor)
+            : this(string.Format("group-{0:X2}", Interlocked.Increment(ref nextId)), executor)
+        {
+        }
+
+        public DefaultChannelGroup(string name, IEventExecutor executor)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException("name");
+            }
+            this.name = name;
+            this.executor = executor;
+        }
+
+        public string Name
+        {
+            get { return this.name; }
+        }
+
+        public IChannel Find(IChannelId id)
+        {
+            IChannel channel = null;
+            if (this.nonServerChannels.TryGetValue(id, out channel))
+            {
+                return channel;
+            }
+            else if (this.serverChannels.TryGetValue(id, out channel))
+            {
+                return channel;
+            }
+            else
+            {
+                return channel;
+            }
+        }
+
+        public Task WriteAsync(object message)
+        {
+            return this.WriteAsync(message, ChannelMatchers.All());
+        }
+
+        public Task WriteAsync(object message, IChannelMatcher matcher)
+        {
+            Contract.Requires(message != null);
+            Contract.Requires(matcher != null);
+            var futures = new Dictionary<IChannel, Task>();
+            foreach (IChannel c in this.nonServerChannels.Values)
+            {
+                if (matcher.Matches(c))
+                {
+                    futures.Add(c, c.WriteAsync(SafeDuplicate(message)));
+                }
+            }
+
+            ReferenceCountUtil.Release(message);
+            return new DefaultChannelGroupCompletionSource(this, futures /*, this.executor*/).Task;
```

Comment:
please file an issue to figure out how to use executor here or what is the right pattern to use.